tests/system/util.py

```diff
@@ -33,21 +47,67 @@ def convert_to_test_response(response: RequestsResponse) -> Response:
 
 class TestVoteService(VoteService):
     url: str
+    datastore: DatastoreService
 
     def vote(self, data: Dict[str, Any]) -> Response:
         ...
 
 
 class TestVoteAdapter(VoteAdapter, TestVoteService):
+    @with_database_context
     def vote(self, data: Dict[str, Any]) -> Response:
         data_copy = copy.deepcopy(data)
+        poll = self.datastore.get(
+            fqid_from_collection_and_id("poll", data["id"]),
+            mapped_fields=["type", "crypt_key", "crypt_signature"],
+            lock_result=False,
+        )
         del data_copy["id"]
+        if poll["type"] == Poll.TYPE_CRYPTOGRAPHIC:
+            crypt_key = base64.b64decode(poll["crypt_key"])
+            crypt_signature = base64.b64decode(poll["crypt_signature"])
+            self.encrypt_votes(data_copy, crypt_key, crypt_signature)
         response = self.make_request(
             self.url.replace("internal", "system") + f"?id={data['id']}",
             data_copy,
         )
         return convert_to_test_response(response)
 
+    def encrypt_votes(
+        self, data: Dict[str, Any], crypt_key: bytes, crypt_signature: bytes
+    ) -> None:
+        pubKeySize = 32
+        nonceSize = 12
+        public_main_key = ed25519.Ed25519PublicKey.from_public_bytes(PUBLIC_MAIN_KEY)
+        public_main_key.verify(crypt_signature, crypt_key)
+
+        private_key = x25519.X25519PrivateKey.generate()
+        public_private_key = private_key.public_key().public_bytes(
+            encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw
+        )
+        public_poll_key = x25519.X25519PublicKey.from_public_bytes(crypt_key)
+        shared_key = private_key.exchange(public_poll_key)
+        derived_key = HKDF(
+            algorithm=hashes.SHA256(),
+            length=pubKeySize,
+            salt=None,
+            info=None,
+        ).derive(shared_key)
+        nonce = os.urandom(nonceSize)
+        cipher = Cipher(algorithms.AES(derived_key), modes.GCM(nonce))
+        encryptor = cipher.encryptor()
+        value_string = str(data.get("value")).replace("'", '"')
+        encrypt_string = bytes(
+            f'{{"votes":{value_string},"token":"12345678"}}', encoding="utf-8"
+        )
+        encrypted = encryptor.update(encrypt_string)
+        encryptor.finalize()
```

Comment:
You have to add the return value of finalize to the encrypted data. In my tests, it always was a empty byte-string. But I don't know if this is always the case. If you don't want to add it, then you should add an assert statement, that it is empty.

The documentation of the function says, that it has to be added. See the example here: https://cryptography.io/en/latest/hazmat/primitives/symmetric-encryption/#cryptography.hazmat.primitives.ciphers.Cipher

https://cryptography.io/en/latest/hazmat/primitives/symmetric-encryption/#cryptography.hazmat.primitives.ciphers.CipherContext
> Once that is done call finalize() to finish the operation and obtain the remainder of the data.